src/api/blob/controller.go

```diff
@@ -0,0 +1,285 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blob
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/garyburd/redigo/redis"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	util "github.com/goharbor/harbor/src/common/utils/redis"
+	ierror "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/pkg/blob"
+)
+
+var (
+	// Ctl is a global blob controller instance
+	Ctl = NewController()
+)
+
+// Controller defines the operations related with blobs
+type Controller interface {
+	// AttachToArtifact attach blobs to manifest.
+	AttachToArtifact(ctx context.Context, params AttachToArtifactParams) error
```

Comment:
Why we defines the param models for each methods? Define the separated params in the method directly is more easier to use and understand?

Reply:
// AttachToProject attach blob to project,
	// `BlobID` in params will used to attach to the project when it provided,
	// otherwise the blob query by digest will be attached.
	AttachToProject(ctx context.Context, params AttachToProjectParams) error

Define the separated params in the method directly means there are two different methods for `AttachToProject`.

1. AttachToProjectByID(ctx context.Context, blobID int64, projectID int64) error

2. AttachToProjectByDigest(ctx context.Context, blobDigest string, projectID int64) error

Is that right?

Reply:
Also four methods for each `Exist` and `Get` methods.

Reply:
Other methods were changed to use params directly, and add options to `Get`  and `Exist` methods to replace the params.